File: internal/game/physics3d.go
package game

import "math"

// Physics uses map pixels, with X/Y horizontal and Z up. Render adapters map
// this to Three.js (X/16, Z/16, Y/16). Positions are at the character's feet.
// Snappy jump tuning: apex ~36px (one body height) in ~0.55s. The client's
// TypeScript port (src/three/physics3d.ts) mirrors these — change together.
const Gravity3D = 960.0
const JumpVelocity3D = 264.0

type Vec3 struct{ X, Y, Z float64 }
type AABB3D struct{ Min, Max Vec3 }
type Body3D struct {
	Position, Velocity Vec3
	Radius, Height     float64
	Grounded           bool
}

// PhysicsWorld3D is immutable during a simulation pass. Both regular maps and
// private camp instances use this same swept upright-cylinder controller.
type PhysicsWorld3D struct {
	Bounds    AABB3D
	HeightAt  func(x, y float64) float64
	Colliders []AABB3D
	// NearbyColliders supplies legacy map collision without scanning the world.
	NearbyColliders               func(minX, minY, maxX, maxY float64) []AABB3D
	Gravity, StepHeight, MaxSlope float64
}

func NewCharacterBody3D(x, y, z float64) Body3D {
	return Body3D{Position: Vec3{x, y, z}, Radius: PlayerCollisionRadius, Height: 36}
}

func finite3(v float64) bool { return !math.IsNaN(v) && !math.IsInf(v, 0) }
func (w PhysicsWorld3D) height(x, y float64) float64 {
	if w.HeightAt == nil {
		return 0
	}
	return w.HeightAt(x, y)
}
func (w PhysicsWorld3D) nearby(x, y, r float64) []AABB3D {
	if w.NearbyColliders == nil {
		return w.Colliders
	}
	return append(w.NearbyColliders(x-r, y-r, x+r, y+r), w.Colliders...)
}
func overlapsDisk(x, y, r float64, b AABB3D) bool {
	dx := x - math.Max(b.Min.X, math.Min(x, b.Max.X))
	dy := y - math.Max(b.Min.Y, math.Min(y, b.Max.Y))
	return dx*dx+dy*dy < r*r-1e-8
}
func (w PhysicsWorld3D) blocked(b Body3D, p Vec3) bool {
	if w.Bounds.Max.X > w.Bounds.Min.X && (p.X-b.Radius < w.Bounds.Min.X || p.X+b.Radius > w.Bounds.Max.X) {
		return true
	}
	if w.Bounds.Max.Y > w.Bounds.Min.Y && (p.Y-b.Radius < w.Bounds.Min.Y || p.Y+b.Radius > w.Bounds.Max.Y) {
		return true
	}
	for _, a := range w.nearby(p.X, p.Y, b.Radius) {
		if p.Z < a.Max.Z-1e-6 && p.Z+b.Height > a.Min.Z+1e-6 && overlapsDisk(p.X, p.Y, b.Radius, a) {
			return true
		}
	}
	return false
}

// support finds the highest surface at or below limit. It permits stepping
// onto low platforms while tall obstacles remain solid and roofs stay above.
func (w PhysicsWorld3D) support(b Body3D, x, y, limit float64) float64 {
	h := w.height(x, y)
	for _, a := range w.nearby(x, y, b.Radius) {
		if a.Max.Z <= limit+1e-6 && a.Max.Z > h && overlapsDisk(x, y, b.Radius, a) {
			h = a.Max.Z
		}
	}
	return h
}

// depenetrate pushes a body out of any collider it overlaps — without it a
// body seated inside a solid box (e.g. furniture placed over the camp entry)
// can never move: every swept candidate stays blocked. The disk exits through
// the nearest face; bounded passes resolve corner pockets without looping.
func (w PhysicsWorld3D) depenetrate(b *Body3D) {
	if w.Bounds.Max.X > w.Bounds.Min.X {
		b.Position.X = math.Max(w.Bounds.Min.X+b.Radius, math.Min(w.Bounds.Max.X-b.Radius, b.Position.X))
	}
	if w.Bounds.Max.Y > w.Bounds.Min.Y {
		b.Position.Y = math.Max(w.Bounds.Min.Y+b.Radius, math.Min(w.Bounds.Max.Y-b.Radius, b.Position.Y))
	}
	for pass := 0; pass < 3; pass++ {
		moved := false
		for _, a := range w.nearby(b.Position.X, b.Position.Y, b.Radius) {
			if !(b.Position.Z < a.Max.Z-1e-6 && b.Position.Z+b.Height > a.Min.Z+1e-6) {
				continue
			}
			cx := math.Max(a.Min.X, math.Min(b.Position.X, a.Max.X))
			cy := math.Max(a.Min.Y, math.Min(b.Position.Y, a.Max.Y))
			dx, dy := b.Position.X-cx, b.Position.Y-cy
			d2 := dx*dx + dy*dy
			if d2 >= b.Radius*b.Radius-1e-8 {
				continue
			}
			if d2 > 1e-10 {
				// Center outside the box — push radially to the surface.
				d := math.Sqrt(d2)
				push := (b.Radius-d)/d + 1e-6
				b.Position.X += dx * push
				b.Position.Y += dy * push
			} else {
				// Center inside the box — exit through the nearest face.
				best := math.Inf(1)
				bx, by := 0.0, 0.0
				for _, e := range [][3]float64{
					{b.Position.X - (a.Min.X - b.Radius), -1, 0},
					{(a.Max.X + b.Radius) - b.Position.X, 1, 0},
					{b.Position.Y - (a.Min.Y - b.Radius), 0, -1},
					{(a.Max.Y + b.Radius) - b.Position.Y, 0, 1},
				} {
					if e[0] > 0 && e[0] < best {
						best, bx, by = e[0], e[1], e[2]
					}
				}
				if math.IsInf(best, 1) {
					continue
				}
				b.Position.X += bx * best
				b.Position.Y += by * best
			}
			moved = true
		}
		if !moved {
			break
		}
	}
}

// Move sweeps horizontal motion in increments smaller than the body radius,
// then integrates vertical motion in bounded timesteps. Neither endpoint-only
// collision nor client-provided elevation is used. dt=0 resolves planar intent
// without advancing gravity (the server tick owns physical time).
func (w PhysicsWorld3D) Move(b Body3D, target Vec2, dt float64) Body3D {
	if !finite3(target.X) || !finite3(target.Y) || !finite3(dt) || !finite3(b.Position.Z) {
		return b
	}
	if b.Radius <= 0 {
		b.Radius = PlayerCollisionRadius
	}
	if b.Height <= 0 {
		b.Height = 36
	}
	w.depenetrate(&b)
	dt = math.Max(0, math.Min(.25, dt))
	gravity := w.Gravity
	if gravity <= 0 {
		gravity = Gravity3D
	}
	maxSlope := w.MaxSlope
	if maxSlope <= 0 {
		maxSlope = 1
	}
	dx, dy := target.X-b.Position.X, target.Y-b.Position.Y
	steps := int(math.Ceil(math.Max(math.Hypot(dx, dy)/math.Max(1, b.Radius*.4), dt/.016)))
	if steps < 1 {
		steps = 1
	}
	if steps > 4096 {
		return b
	}
	sx, sy, sd := dx/float64(steps), dy/float64(steps), dt/float64(steps)
	for i := 0; i < steps; i++ {
		from := b.Position
		floor := w.support(b, from.X, from.Y, from.Z+.001)
		if from.Z < floor {
			b.Position.Z = floor
			b.Grounded = true
			b.Velocity.Z = 0
			from = b.Position
		}
		if math.Abs(from.Z-floor) < .001 && b.Velocity.Z <= 0 {
			b.Grounded = true
		}
		try := func(x, y float64) bool {
			p := Vec3{x, y, b.Position.Z}
			ground := w.height(x, y)
			if b.Grounded {
				// Compare actual mesh incline independently of step height, otherwise
				// repeated tiny steps could climb arbitrarily steep terrain.
				delta := ground - w.height(from.X, from.Y)
				if delta > math.Hypot(x-from.X, y-from.Y)*maxSlope+1e-6 {
					return false
				}
				support := w.support(b, x, y, from.Z+w.StepHeight)
				if support > from.Z+w.StepHeight+1e-6 {
					return false
				}
				if support >= from.Z-w.StepHeight {
					p.Z = support
				}
			} else if ground > p.Z {
				return false
			}
			if w.blocked(b, p) {
				return false
			}
			b.Position = p
			return true
		}
		if !try(from.X+sx, from.Y+sy) {
			if !try(from.X+sx, from.Y) {
				try(from.X, from.Y+sy)
			}
		}
		floor = w.support(b, b.Position.X, b.Position.Y, b.Position.Z+.001)
		if b.Position.Z > floor+.001 || b.Velocity.Z > 0 {
			b.Grounded = false
			vz := b.Velocity.Z - gravity*sd
			nextZ := b.Position.Z + (b.Velocity.Z+vz)*.5*sd
			if vz > 0 {
				for _, a := range w.nearby(b.Position.X, b.Position.Y, b.Radius) {
					if overlapsDisk(b.Position.X, b.Position.Y, b.Radius, a) && b.Position.Z+b.Height <= a.Min.Z+1e-6 && nextZ+b.Height >= a.Min.Z {
						nextZ = a.Min.Z - b.Height
						vz = 0
					}
				}
			}
			// Land only when actually descending onto (or resting on) the floor. A
			// rising body with dt=0 — e.g. a jump whose intent resolves planar motion
			// in the same call — computes nextZ==Position.Z==floor and must keep its
			// upward velocity rather than settling instantly.
			if nextZ <= floor {
				nextZ = floor
				if vz <= 0 {
					vz = 0
					b.Grounded = true
				}
			}
			b.Position.Z = nextZ
			b.Velocity.Z = vz
		} else {
			b.Position.Z = floor
			b.Velocity.Z = 0
			b.Grounded = true
		}
	}
	return b
}

// LineOfSight tests full 3D AABB slabs and samples the heightfield at less than
// one map pixel intervals. Broad-phase candidates are bounded by the segment.
func (w PhysicsWorld3D) LineOfSight(from, to Vec3) bool {
	boxes := w.Colliders
	if w.NearbyColliders != nil {
		boxes = append(w.NearbyColliders(math.Min(from.X, to.X), math.Min(from.Y, to.Y), math.Max(from.X, to.X), math.Max(from.Y, to.Y)), boxes...)
	}
	for _, b := range boxes {
		enter, leave := 0.0, 1.0
		for _, axis := range [][4]float64{{from.X, to.X, b.Min.X, b.Max.X}, {from.Y, to.Y, b.Min.Y, b.Max.Y}, {from.Z, to.Z, b.Min.Z, b.Max.Z}} {
			d := axis[1] - axis[0]
			if math.Abs(d) < 1e-9 {
				if axis[0] < axis[2] || axis[0] > axis[3] {
					leave = -1
				}
				continue
			}
			a, z := (axis[2]-axis[0])/d, (axis[3]-axis[0])/d
			if a > z {
				a, z = z, a
			}
			enter = math.Max(enter, a)
			leave = math.Min(leave, z)
		}
		if enter <= leave && leave > 1e-5 && enter < 1-1e-5 {
			return false
		}
	}
	steps := int(math.Ceil(math.Hypot(to.X-from.X, to.Y-from.Y)))
	for i := 1; i < steps; i++ {
		t := float64(i) / float64(steps)
		x, y := from.X+(to.X-from.X)*t, from.Y+(to.Y-from.Y)*t
		if from.Z+(to.Z-from.Z)*t < w.height(x, y) {
			return false
		}
	}
	return true
}

File: wails/frontend/src/three/physics3d.ts
import * as THREE from "three";
import { WorldHeightmap, WORLD_SCALE } from "./heightmap";
import { WALKABLE } from "../world/overworld";
import type { Scene3DDoc, SceneTransform } from "./scene3d";

// Map-pixel 3D physics — a TypeScript port of internal/game/physics3d.go used
// for client-side prediction. Positions are at the character's feet: X/Y are
// horizontal map pixels, Z is up (Three.js Y). Constants must stay in sync
// with the Go solver and hub.go's jump impulse — prediction only stays
// truthful while both sides agree.

export const PLAYER_BODY_RADIUS = 15.625; // game.PlayerCollisionRadius
export const BODY_HEIGHT = 36;
// Snappy jump: apex ~36px (one body height), ~0.55s total. Mirrors
// game.Gravity3D / game.JumpVelocity3D — change together.
export const JUMP_VELOCITY = 264;
export const GRAVITY = 960, STEP_HEIGHT = 10, MAX_SLOPE = 1.2;

export interface Vec3 { x: number; y: number; z: number }
export interface AABB3D { min: Vec3; max: Vec3 }

export interface Body3D {
  pos: Vec3;
  velZ: number;
  grounded: boolean;
  radius: number;
  height: number;
}

export function newBody3D(x: number, y: number, z: number): Body3D {
  return { pos: { x, y, z }, velZ: 0, grounded: true, radius: PLAYER_BODY_RADIUS, height: BODY_HEIGHT };
}

const finite = (v: number) => Number.isFinite(v);

function overlapsDisk(x: number, y: number, r: number, b: AABB3D): boolean {
  const dx = x - Math.max(b.min.x, Math.min(x, b.max.x));
  const dy = y - Math.max(b.min.y, Math.min(y, b.max.y));
  return dx * dx + dy * dy < r * r - 1e-8;
}

/** Swept upright-cylinder controller mirroring game.PhysicsWorld3D. */
export class PhysicsWorld3D {
  bounds?: AABB3D;
  heightAt: (x: number, y: number) => number = () => 0;
  colliders: AABB3D[] = [];
  /** Supplies legacy tile collision without scanning the whole world. */
  nearbyColliders?: (minX: number, minY: number, maxX: number, maxY: number) => AABB3D[];
  gravity = GRAVITY;
  stepHeight = STEP_HEIGHT;
  maxSlope = MAX_SLOPE;

  private nearby(x: number, y: number, r: number): AABB3D[] {
    if (!this.nearbyColliders) return this.colliders;
    return [...this.nearbyColliders(x - r, y - r, x + r, y + r), ...this.colliders];
  }

  private blocked(b: Body3D, p: Vec3): boolean {
    const bounds = this.bounds;
    if (bounds) {
      if (bounds.max.x > bounds.min.x && (p.x - b.radius < bounds.min.x || p.x + b.radius > bounds.max.x)) return true;
      if (bounds.max.y > bounds.min.y && (p.y - b.radius < bounds.min.y || p.y + b.radius > bounds.max.y)) return true;
    }
    for (const a of this.nearby(p.x, p.y, b.radius)) {
      if (p.z < a.max.z - 1e-6 && p.z + b.height > a.min.z + 1e-6 && overlapsDisk(p.x, p.y, b.radius, a)) return true;
    }
    return false;
  }

  /** Highest surface at or below limit — permits stepping onto low platforms. */
  private support(b: Body3D, x: number, y: number, limit: number): number {
    let h = this.heightAt(x, y);
    for (const a of this.nearby(x, y, b.radius)) {
      if (a.max.z <= limit + 1e-6 && a.max.z > h && overlapsDisk(x, y, b.radius, a)) h = a.max.z;
    }
    return h;
  }

  /**
   * Sweeps horizontal motion in increments smaller than the body radius, then
   * integrates vertical motion in bounded timesteps. dt=0 resolves planar
   * intent without advancing gravity. Mirrors game.PhysicsWorld3D.Move.
   */
  move(b: Body3D, tx: number, ty: number, dt: number): Body3D {
    if (!finite(tx) || !finite(ty) || !finite(dt) || !finite(b.pos.z)) return b;
    if (b.radius <= 0) b.radius = PLAYER_BODY_RADIUS;
    if (b.height <= 0) b.height = BODY_HEIGHT;
    dt = Math.max(0, Math.min(.25, dt));
    const gravity = this.gravity > 0 ? this.gravity : GRAVITY;
    const maxSlope = this.maxSlope > 0 ? this.maxSlope : 1;
    const dx = tx - b.pos.x, dy = ty - b.pos.y;
    let steps = Math.ceil(Math.max(Math.hypot(dx, dy) / Math.max(1, b.radius * .4), dt / .016));
    if (steps < 1) steps = 1;
    if (steps > 4096) return b;
    const sx = dx / steps, sy = dy / steps, sd = dt / steps;
    for (let i = 0; i < steps; i++) {
      let from = { ...b.pos };
      let floor = this.support(b, from.x, from.y, from.z + .001);
      if (from.z < floor) {
        b.pos.z = floor; b.grounded = true; b.velZ = 0;
        from = { ...b.pos };
      }
      if (Math.abs(from.z - floor) < .001 && b.velZ <= 0) b.grounded = true;
      const tryStep = (x: number, y: number): boolean => {
        const p: Vec3 = { x, y, z: b.pos.z };
        const ground = this.heightAt(x, y);
        if (b.grounded) {
          // Compare actual mesh incline independently of step height, otherwise
          // repeated tiny steps could climb arbitrarily steep terrain.
          const delta = ground - this.heightAt(from.x, from.y);
          if (delta > Math.hypot(x - from.x, y - from.y) * maxSlope + 1e-6) return false;
          const sup = this.support(b, x, y, from.z + this.stepHeight);
          if (sup > from.z + this.stepHeight + 1e-6) return false;
          if (sup >= from.z - this.stepHeight) p.z = sup;
        } else if (ground > p.z) {
          return false;
        }
        if (this.blocked(b, p)) return false;
        b.pos = p;
        return true;
      };
      if (!tryStep(from.x + sx, from.y + sy)) {
        if (!tryStep(from.x + sx, from.y)) tryStep(from.x, from.y + sy);
      }
      floor = this.support(b, b.pos.x, b.pos.y, b.pos.z + .001);
      if (b.pos.z > floor + .001 || b.velZ > 0) {
        b.grounded = false;
        let vz = b.velZ - gravity * sd;
        let nextZ = b.pos.z + (b.velZ + vz) * .5 * sd;
        if (vz > 0) {
          for (const a of this.nearby(b.pos.x, b.pos.y, b.radius)) {
            if (overlapsDisk(b.pos.x, b.pos.y, b.radius, a) && b.pos.z + b.height <= a.min.z + 1e-6 && nextZ + b.height >= a.min.z) {
              nextZ = a.min.z - b.height; vz = 0;
            }
          }
        }
        // Land only when actually descending onto (or resting on) the floor — a
        // rising body with dt=0 must keep its upward velocity.
        if (nextZ <= floor) {
          nextZ = floor;
          if (vz <= 0) { vz = 0; b.grounded = true; }
        }
        b.pos.z = nextZ;
        b.velZ = vz;
      } else {
        b.pos.z = floor; b.velZ = 0; b.grounded = true;
      }
    }
    return b;
  }
}

const DEG = Math.PI / 180;

function localSceneMatrix(t: SceneTransform): THREE.Matrix4 {
  return new THREE.Matrix4().compose(
    new THREE.Vector3(t.position[0], t.position[1], t.position[2]),
    new THREE.Quaternion().setFromEuler(new THREE.Euler(t.rotation[0] * DEG, t.rotation[1] * DEG, t.rotation[2] * DEG)),
    new THREE.Vector3(t.scale[0], t.scale[1], t.scale[2]),
  );
}

/**
 * Authored collider components as physics-space AABBs — mirrors
 * game.SceneColliders3D: the 8 offset±size/2 corners go through the object's
 * world matrix, then Three.js world units convert to map pixels with the Y/Z
 * axis swap (world x→x, world z→y, world y→z-up).
 */
export function sceneColliders3D(doc: Scene3DDoc): AABB3D[] {
  const byId = new Map(doc.objects.map(o => [o.id, o]));
  const cache = new Map<string, THREE.Matrix4>();
  const world = (id: string, depth: number): THREE.Matrix4 => {
    const cached = cache.get(id);
    if (cached) return cached;
    const o = byId.get(id);
    const m = !o || depth > doc.objects.length
      ? new THREE.Matrix4()
      : (o.parent ? world(o.parent, depth + 1) : new THREE.Matrix4()).multiply(localSceneMatrix(o.transform));
    cache.set(id, m);
    return m;
  };
  const out: AABB3D[] = [];
  const corner = new THREE.Vector3();
  for (const o of doc.objects) {
    const c = o.components?.collider;
    if (!c || !c.enabled || c.isTrigger) continue;
    const m = world(o.id, 0);
    const box: AABB3D = { min: { x: Infinity, y: Infinity, z: Infinity }, max: { x: -Infinity, y: -Infinity, z: -Infinity } };
    for (let i = 0; i < 8; i++) {
      corner.set(
        c.offset[0] + (i & 1 ? 1 : -1) * c.size[0] / 2,
        c.offset[1] + (i & 2 ? 1 : -1) * c.size[1] / 2,
        c.offset[2] + (i & 4 ? 1 : -1) * c.size[2] / 2,
      ).applyMatrix4(m);
      const px = corner.x / WORLD_SCALE, py = corner.z / WORLD_SCALE, pz = corner.y / WORLD_SCALE;
      if (px < box.min.x) box.min.x = px;
      if (py < box.min.y) box.min.y = py;
      if (pz < box.min.z) box.min.z = pz;
      if (px > box.max.x) box.max.x = px;
      if (py > box.max.y) box.max.y = py;
      if (pz > box.max.z) box.max.z = pz;
    }
    out.push(box);
  }
  return out;
}

/**
 * Predicted physics world matching Overworld.Physics3D(): map bounds, the
 * heightmap as HeightAt (WorldHeightmap already mirrors TerrainHeightAt),
 * authored scene colliders, and one AABB per unwalkable tile. Water cells top
 * out at h+8 (wading depth), everything else solid to h+48.
 */
export function overworldPhysics3D(field: WorldHeightmap, doc?: Scene3DDoc | null): PhysicsWorld3D {
  const w = new PhysicsWorld3D();
  const map = field.map, ts = map.tile;
  w.bounds = { min: { x: 0, y: 0, z: -100000 }, max: { x: map.cols * ts, y: map.rows * ts, z: 100000 } };
  w.heightAt = (x, y) => field.height(x, y) / WORLD_SCALE;
  w.colliders = doc ? sceneColliders3D(doc) : [];
  w.nearbyColliders = (minX, minY, maxX, maxY) => {
    const boxes: AABB3D[] = [];
    const r1 = Math.min(map.rows - 1, Math.floor(maxY / ts));
    const c1 = Math.min(map.cols - 1, Math.floor(maxX / ts));
    for (let r = Math.max(0, Math.floor(minY / ts)); r <= r1; r++) {
      for (let c = Math.max(0, Math.floor(minX / ts)); c <= c1; c++) {
        const cell = field.collisionCell(c, r);
        if (WALKABLE.has(cell)) continue;
        const h = w.heightAt((c + .5) * ts, (r + .5) * ts);
        boxes.push({ min: { x: c * ts, y: r * ts, z: h - 10000 }, max: { x: (c + 1) * ts, y: (r + 1) * ts, z: h + (cell === "~" ? 8 : 48) } });
      }
    }
    return boxes;
  };
  return w;
}

export type ReconcileVerdict = "snap" | "adjust" | "ok";

/**
 * Merge an authoritative entity update into the predicted body. Large planar
 * or vertical divergence snaps outright (teleport/knockback); small vertical
 * drift is absorbed when both sides agree the body is grounded, and an
 * authoritative landing sticks when the prediction is already descending.
 */
export function reconcileBody3D(
  body: Body3D,
  auth: { x: number; y: number; z: number; grounded?: boolean },
): ReconcileVerdict {
  const planar = Math.hypot(auth.x - body.pos.x, auth.y - body.pos.y);
  const dz = auth.z - body.pos.z;
  if (planar > 80 || Math.abs(dz) > 64 || !finite(body.pos.z)) {
    body.pos.x = auth.x; body.pos.y = auth.y; body.pos.z = auth.z;
    body.velZ = 0;
    body.grounded = auth.grounded ?? true;
    return "snap";
  }
  if (auth.grounded) {
    if (body.grounded) {
      body.pos.z = auth.z;
      return "adjust";
    }
    if (body.velZ <= 0 && body.pos.z <= auth.z + 4) {
      body.pos.z = auth.z; body.velZ = 0; body.grounded = true;
      return "adjust";
    }
  }
  return "ok";
}
